executor/runcexecutor/executor_linux.go

```diff
@@ -0,0 +1,308 @@
+package runcexecutor
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"os/signal"
+	"sync"
+	"time"
+
+	"github.com/containerd/console"
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/cio"
+	runcV2 "github.com/containerd/containerd/runtime/v2/runc"
+	"github.com/containerd/containerd/sys/reaper"
+	runc "github.com/containerd/go-runc"
+	"github.com/moby/buildkit/executor"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sys/unix"
+)
+
+var setupReaper sync.Once
+
+func handleSignals(ctx context.Context) error {
+	signals := make(chan os.Signal, 32)
+	signal.Notify(signals, unix.SIGPIPE, unix.SIGCHLD)
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case s := <-signals:
+			switch s {
+			case unix.SIGCHLD:
+				if err := reaper.Reap(); err != nil {
+					logrus.WithError(err).Error("reap exit status")
+				}
+			case unix.SIGPIPE:
+			}
+		}
+	}
+}
+
+func (w *runcExecutor) initPlatform() (*runcExecutor, error) {
+	setupReaper.Do(func() {
+		runc.Monitor = reaper.Default
+		// collect sigchld from grandchildren processes so we can see when
+		// the forked/detached runc processes exit
+		reaper.SetSubreaper(1)
+		go handleSignals(context.Background())
+	})
+
+	stdioPlatform, err := runcV2.NewPlatform()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to initialize stdio platform")
+	}
+	w.stdioPlatform = stdioPlatform
+	return w, nil
+}
+
+func (w *runcExecutor) run(ctx context.Context, id, bundle string, process executor.ProcessInfo, ended chan struct{}) (int, error) {
+	var endedOnce sync.Once
+	defer endedOnce.Do(func() {
+		close(ended)
+	})
+	pidfile, err := ioutil.TempFile(bundle, "run-*.pid")
+	if err != nil {
+		return containerd.UnknownExitStatus, errors.Wrap(err, "failed to create pidfile")
+	}
+	pidfile.Close()
```

Comment:
Defer remove?

Reply:
It should be unnecessary for pid1 since we `os.RemoveAll` for the entire `bundle` directory when `(*runcExecutor).Run` ends.